Hazelcast terminates listener incorrectly.

I have this Hazelcast set up

Application ->hzClient  (3.7.5 version)    ------------------- hzServer1 (3.7.5)
                                                                 ------------------- hzServer2 (3.7.5)
hzServer1 and hzServer2 are members in the same cluster.
hzClient registers message listener with the cluster to receive messages from reliable topic.


If I restart hzServer2 then I see the exception below and the listener is terminated incorrectly. There is no call back to tell application that the listener is terminated.
I end up with situation where application no longer receives any message from the topic.

I have the following suggestions.
1) ReliableMessageListener should not be terminated on exception. If exception occurs when trying to read message from the dead member, don't terminate the listener, try re-creating the listener on different member. If there is no alive member, wait for the connection to re-established (could be just a temporary glitch) then re-create listeners again.
2) Look like there is existing isTerminal(Throwable failure). Should that be used to feed the exception to application? 
3) If the listener has to be terminated (no other choice), there should be a new method introduced into ReliableMessageListener interface (ex: listenerTerminated(...)) to tell application that the listener is terminated. 
Terminate listener silently without telling application doesn't make any sense at all because application assumes everything is well when it is not.

-------------------------------------------------------------------------------------------------------------------

```
        hz.client_0 [hzc1] [3.7.5] Terminating MessageListener com.broadsoft.persistence.hazelcast.BWHazelcastInstance$TopicContainer$ReliableMessageListenerImpl@1f739e95 on topic: profileManagementUpdate. Reason: Unhandled exception, message: 
                                          com.hazelcast.spi.exception.TargetDisconnectedException: Disconnecting from member [10.16.176.21]:5701 due to heartbeat problems. Current time: 2017-01-26 12:50:57.655. Last heartbeat requested: 2017-01-26 12:50:55.660. Last heartbeat received: 2017-01-26 12:50:55.662. Last read: 2017-01-26 12:50:57.137. Connection ClientConnection{live=false, connectionId=2, socketChannel=SSLSocketChannelWrapper{socketChannel=java.nio.channels.SocketChannel[closed]}, remoteEndpoint=[10.16.176.21]:5701, lastReadTime=2017-01-26 12:50:57.137, lastWriteTime=2017-01-26 12:50:55.661, closedTime=2017-01-26 12:50:57.137, lastHeartbeatRequested=2017-01-26 12:50:55.660, lastHeartbeatReceived=2017-01-26 12:50:55.662, connected server version=3.7.5} 
<java.util.concurrent.ExecutionException: com.hazelcast.spi.exception.TargetDisconnectedException: Disconnecting from member [10.16.176.21]:5701 due to heartbeat problems. Current time: 2017-01-26 12:50:57.655. Last heartbeat requested: 2017-01-26 12:50:55.660. Last heartbeat received: 2017-01-26 12:50:55.662. Last read: 2017-01-26 12:50:57.137. Connection ClientConnection{live=false, connectionId=2, socketChannel=SSLSocketChannelWrapper{socketChannel=java.nio.channels.SocketChannel[closed]}, remoteEndpoint=[10.16.176.21]:5701, lastReadTime=2017-01-26 12:50:57.137, lastWriteTime=2017-01-26 12:50:55.661, closedTime=2017-01-26 12:50:57.137, lastHeartbeatRequested=2017-01-26 12:50:55.660, lastHeartbeatReceived=2017-01-26 12:50:55.662, connected server version=3.7.5}>
 java.util.concurrent.ExecutionException: com.hazelcast.spi.exception.TargetDisconnectedException: Disconnecting from member [10.16.176.21]:5701 due to heartbeat problems. Current time: 2017-01-26 12:50:57.655. Last heartbeat requested: 2017-01-26 12:50:55.660. Last heartbeat received: 2017-01-26 12:50:55.662. Last read: 2017-01-26 12:50:57.137. Connection ClientConnection{live=false, connectionId=2, socketChannel=SSLSocketChannelWrapper{socketChannel=java.nio.channels.SocketChannel[closed]}, remoteEndpoint=[10.16.176.21]:5701, lastReadTime=2017-01-26 12:50:57.137, lastWriteTime=2017-01-26 12:50:55.661, closedTime=2017-01-26 12:50:57.137, lastHeartbeatRequested=2017-01-26 12:50:55.660, lastHeartbeatReceived=2017-01-26 12:50:55.662, connected server version=3.7.5}
        at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolve(ClientInvocationFuture.java:66)
        at com.hazelcast.spi.impl.AbstractInvocationFuture$1.run(AbstractInvocationFuture.java:246)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
        at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
        at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:92)
Caused by: com.hazelcast.spi.exception.TargetDisconnectedException: Disconnecting from member [10.16.176.21]:5701 due to heartbeat problems. Current time: 2017-01-26 12:50:57.655. Last heartbeat requested: 2017-01-26 12:50:55.660. Last heartbeat received: 2017-01-26 12:50:55.662. Last read: 2017-01-26 12:50:57.137. Connection ClientConnection{live=false, connectionId=2, socketChannel=SSLSocketChannelWrapper{socketChannel=java.nio.channels.SocketChannel[closed]}, remoteEndpoint=[10.16.176.21]:5701, lastReadTime=2017-01-26 12:50:57.137, lastWriteTime=2017-01-26 12:50:55.661, closedTime=2017-01-26 12:50:57.137, lastHeartbeatRequested=2017-01-26 12:50:55.660, lastHeartbeatReceived=2017-01-26 12:50:55.662, connected server version=3.7.5}
        at com.hazelcast.spi.exception.TargetDisconnectedException.newTargetDisconnectedExceptionCausedByHeartbeat(TargetDisconnectedException.java:66)
        at com.hazelcast.client.spi.impl.ClientInvocationServiceSupport$CleanResourcesTask.run(ClientInvocationServiceSupport.java:221)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
        ... 5 more
Caused by: java.io.EOFException: Remote socket closed!
        at com.hazelcast.client.connection.nio.ClientReadHandler.handle(ClientReadHandler.java:87)
        at com.hazelcast.nio.tcp.nonblocking.NonBlockingIOThread.handleSelectionKey(NonBlockingIOThread.java:345)
        at com.hazelcast.nio.tcp.nonblocking.NonBlockingIOThread.handleSelectionKeys(NonBlockingIOThread.java:330)
        at com.hazelcast.nio.tcp.nonblocking.NonBlockingIOThread.selectLoop(NonBlockingIOThread.java:248)
        at com.hazelcast.nio.tcp.nonblocking.NonBlockingIOThread.run(NonBlockingIOThread.java:201)
```

--- a/hazelcast-client/src/test/java/com/hazelcast/client/topic/Issue9766Test.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/topic/Issue9766Test.java
@@ -1,0 +1,85 @@
+package com.hazelcast.client.topic;/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import com.hazelcast.client.HazelcastClient;
+import com.hazelcast.client.test.TestHazelcastFactory;
+import com.hazelcast.config.Config;
+import com.hazelcast.config.RingbufferConfig;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.ITopic;
+import com.hazelcast.core.Message;
+import com.hazelcast.core.MessageListener;
+import com.hazelcast.instance.HazelcastInstanceImpl;
+import com.hazelcast.test.HazelcastParallelClassRunner;
+import com.hazelcast.test.annotation.ParallelTest;
+import com.hazelcast.test.annotation.QuickTest;
+import com.hazelcast.topic.ReliableMessageListener;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import java.util.concurrent.CountDownLatch;
+
+import static com.hazelcast.test.HazelcastTestSupport.assertOpenEventually;
+
+/**
+ * Test for #9766 (https://github.com/hazelcast/hazelcast/issues/9766)
+ */
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelTest.class})
+
+public class Issue9766Test {
+    private final TestHazelcastFactory hazelcastFactory = new TestHazelcastFactory();
+
+    static final String topicName = "foo";
+
+    @After
+    public void tearDown() {
+        hazelcastFactory.terminateAll();
+    }
+
+    @Test
+    public void serverRestartWhenReliableTopicListenerRegistered() {
+        HazelcastInstance server = hazelcastFactory.newHazelcastInstance();
+
+        HazelcastInstance hazelcastClient = hazelcastFactory.newHazelcastClient();
+        HazelcastInstance hazelcastClient2 = hazelcastFactory.newHazelcastClient();
+        ITopic<Integer> topic = hazelcastClient.getReliableTopic(topicName);
+        final ITopic<Integer> topic2 = hazelcastClient2.getReliableTopic(topicName);
+
+        final CountDownLatch listenerLatch = new CountDownLatch(1);
+
+        // Add listener using the first client
+        topic.addMessageListener(new MessageListener<Integer>() {
+            @Override
+            public void onMessage(Message<Integer> message) {
+                listenerLatch.countDown();
+            }
+        });
+
+        // restart the server
+        server.getLifecycleService().terminate();
+
+        hazelcastFactory.newHazelcastInstance();
+
+        // publish some data
+        topic2.publish(5);
+
+        assertOpenEventually(listenerLatch);
+    }
+}

--- a/pom.xml
+++ b/pom.xml
@@ -38,7 +38,7 @@
 
     <properties>
         <main.basedir>${project.basedir}</main.basedir>
-        <client.protocol.version>1.4.0-11</client.protocol.version>
+        <client.protocol.version>1.4.0-12</client.protocol.version>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
 
         <jdk.version>1.6</jdk.version>

